Testnet/NFTAuctionStore/NFTAuctionStore/NFTAuctionStore.cs

```diff
@@ -0,0 +1,199 @@
+﻿using Stratis.SmartContracts;
+
+[Deploy]
+public class NFTAuctionStore : SmartContract
+{
+    private void SetAuctionInfo(Address contract, ulong tokenId, AuctionInfo auctionInfo)
+    {
+        State.SetStruct($"AuctionInfo:{contract}:{tokenId}", auctionInfo);
+    }
+
+    public AuctionInfo GetAuctionInfo(Address contract, ulong tokenId)
+    {
+        return State.GetStruct<AuctionInfo>($"AuctionInfo:{contract}:{tokenId}");
+    }
+
+    public ulong GetRefund(Address address)
+    {
+        return State.GetUInt64($"Refund:{address}");
+    }
+
+    private void SetRefund(Address address, ulong balance)
+    {
+        State.SetUInt64($"Refund:{address}", balance);
+    }
+
+    public NFTAuctionStore(ISmartContractState state)
+        : base(state)
+    {
+        EnsureNotPayable();
+    }
+
+    public void Auction(Address contract, ulong tokenId, ulong startingPrice, ulong duration)
+    {
+        EnsureNotPayable();
+
+        var tokenOwner = GetOwner(contract, tokenId);
+
+        Assert(tokenOwner != Address, "The token is already on sale.");
+
+        EnsureCallerCanOperate(contract, tokenOwner);
+
+        TransferToken(contract, tokenId, tokenOwner, Address);
+
+        var auction = new AuctionInfo
+        {
+            Seller = tokenOwner,
+            EndBlock = checked(Block.Number + duration),
+            StartingPrice = startingPrice,
+            Ended = false
+        };
+
+        SetAuctionInfo(contract, tokenId, auction);
+        Log(new AuctionStartedLog { Contract = contract, TokenId = tokenId, EndBlock = auction.EndBlock, Seller = auction.Seller, startingPrice = auction.StartingPrice });
+    }
+
+    public void Bid(Address contract, ulong tokenId)
+    {
+        var auction = GetAuctionInfo(contract, tokenId);
+
+        Assert(Block.Number < auction.EndBlock, "Auction ended.");
+
+        Assert(Message.Value > auction.HighestBid && Message.Value >= auction.StartingPrice, "The amount is not higher than highest bidder or starting price.");
+
+        if (auction.HighestBid > 0)
+        {
+            //refund for previous bidder
+            var balance = GetRefund(auction.HighestBidder);
+
+            SetRefund(auction.HighestBidder, balance + auction.HighestBid);
+        }
+
+        auction.HighestBidder = Message.Sender;
+        auction.HighestBid = Message.Value;
+
+        SetAuctionInfo(contract, tokenId, auction);
+
+        Log(new HighestBidUpdatedLog { Contract = contract, TokenId = tokenId, Bidder = auction.HighestBidder, Bid = auction.HighestBid });
+    }
+
+    public bool Refund()
+    {
+        EnsureNotPayable();
+
+        var amount = GetRefund(Message.Sender);
+
+        Assert(amount > 0);
+
+        SetRefund(Message.Sender, 0);
+
+        var transfer = Transfer(this.Message.Sender, amount);
+
+        if (!transfer.Success)
+            SetRefund(Message.Sender, amount);
+
```

Comment:
Will this really ever fail? 

Really, this would need to check `refundAmount <= Balance` to check the balance of the contract.

Ultimately, I don't see a need for this public method at all. If a someone bids higher than my previous bid, they should also have to pay the gas to refund my previous bid.

Reply:
This is just safe by default approach. Caller can be another contract and may reject transfer. So it is there for worst case scenario.

Reply:
> Ultimately, I don't see a need for this public method at all. If a someone bids higher than my previous bid, they should also have to pay the gas to refund my previous bid.

I am going to propose another pr for this case.